patterns/strategy.md

```diff
@@ -0,0 +1,175 @@
+
+```rust
+use std::collections::HashMap;
+type Data = HashMap<String, u32>;
+
+trait Formatter {
+    fn format(&self, data: &Data) -> String;
+}
+
+struct Report;
+
+impl Report {
+    fn generate<T: Formatter>(g: T) -> String {
+        // backend operations...
+        let mut data = HashMap::new();
+        data.insert("one".to_string(), 1);
+        data.insert("two".to_string(), 2);
+        // generate report
+        g.format(&data)
+    }
+}
+
+struct Text;
+impl Formatter for Text {
+    fn format(&self, data: &Data) -> String {
+        data.iter()
+            .map(|(key, val)| format!("{} {}\n", key, val))
+            .collect()
+    }
+}
+
+struct Json;
+impl Formatter for Json {
+    fn format(&self, data: &Data) -> String {
+        let mut s = String::from("[");
+        let mut iter = data.into_iter();
+        if let Some((key, val)) = iter.next() {
+            let entry = format!(r#"{{"{}":"{}"}}"#, key, val);
+            s.push_str(&entry);
+            while let Some((key, val)) = iter.next() {
+                s.push(',');
+                let entry = format!(r#"{{"{}":"{}"}}"#, key, val);
+                s.push_str(&entry);
+            }
+        }
+        s.push(']');
+        s
+    }
+}
+
+fn main() {
+    let s = Report::generate(Text);
+    assert!(s.contains("one 1"));
+    assert!(s.contains("two 2"));
+
+    let s = Report::generate(Json);
+    assert!(s.contains(r#"{"one":"1"}"#));
+    assert!(s.contains(r#"{"two":"2"}"#));
+}
+```
+
+## Advantages
+
+The main advantage is separation of concerns. For example, in this case `Report` does not know anything about specific
+implementations of `Json` and `Text`, whereas the output implementations does not care about how data is
+preprocessed, stored, and fetched.
+The only thing they have to know is context and a specific trait and method to implement,
+i.e,`Formatter` and `run`.
+
+## Disadvantages
+
+For each strategy there must be implemented at least one module, so number of modules
+increases with number of strategies.
+If there are many strategies to choose from then users have to know how strategies differ
+from one another.
+
+## Discussion
+
+In the previous example all strategies are implemented in a single file.
+Ways of providing different strategies includes:
+
+- All in one file (as shown in this example, similar to being separated as modules)
+- Separated as modules, E.g. `formatter::json` module, `formatter::text` module
+- Use compiler feature flags, E.g. `json` feature, `text` feature
+- Separated as crates, E.g. `json` crate, `text` crate
+
+Serde crate is a good example of the `Strategy` pattern in action. Serde allows [full customization](https://serde.rs/custom-serialization.html)
+of the serialization behavior by manually implementing `Serialize` and `Deserialize` traits for our type.
+For example, we could easily swap `serde_json` with `serde_cbor` since they expose a similar method.
```

Comment:
Sorry, grammatical error I put just now.
```suggestion
For example, we could easily swap `serde_json` with `serde_cbor` since they expose similar methods.
```